nomenclature/testing.py

```diff
@@ -144,12 +145,16 @@ def assert_valid_structure(
             f"Definitions directory not found: {path / definitions}"
         )
 
-    if dimensions == ():  # if "dimensions" were not specified
-        dimensions = [x.stem for x in (path / definitions).iterdir() if x.is_dir()]
+    if not dimensions:  # if "dimensions" were not specified
+        if path / "nomenclature.yaml" in path.iterdir():
+            dimensions = NomenclatureConfig.from_file(
+                path / "nomenclature.yaml"
+            ).dimensions
         if not dimensions:
-            raise FileNotFoundError(
-                f"`definitions` directory is empty: {path / definitions}"
-            )
+            dimensions = [x.stem for x in (path / definitions).iterdir() if x.is_dir()]
+        if not dimensions:
+            raise FileNotFoundError(f"`definitions` directory is empty: {definitions}")
```

Comment:
All of this is basically already implemented directly within the **DataStructureDefinition**-initialiation, see https://github.com/IAMconsortium/nomenclature/blob/0ae378f418a260315f88c7e19a204e83a8416bcb/nomenclature/definition.py#L66

Reply:
This is what I was pointing out initially about having to set dimensions outside the `DataStructureDefinition`. In order to have the order be:

>    * explicit `dimensions` argument (this is currently tested in `test_cli_custom_dimensions_runs()`)
>    * if not given, use `dimensions` argument set from `nomenclature.yaml` (not sure if this is tested)
>    * if also not given, check all available folders

I define it outside DSD, since within it, the order is: explicit `dimensions` -> `nomenclature.yaml` -> `["region", "variable"]`. The only way (besides changing the code in DSD) to enforce the last step is by doing the order before calling DSD and passing the dimensions. In its current implementation, the DSD won't check folders (hence why `validate-project` does that check in the CLI function too).

Reply:
Ok, I think this is a misunderstanding - this function should just check whether a project can be instantiated correctly, and I think the extra code here is just from long ago when DataStructureDefinition wasn't as elegant as it is now.

Reply:
I think we should go the other way and change the default-dimensions of a DataStructureDefinition to all sub-folders in `definitions`, not hard-coded to region and variable.

But this should be left to another PR, simply revert the changes here and we merge this PR?

Reply:
Agreed that it makes more sense to change the default dimensions in DSD! Will do the suggested changes.